Tests fail if `postgres` extra is missing

```python traceback
$ poetry run pytest tests/test_task_api.py 

====================================================================================================== test session starts ======================================================================================================
platform linux -- Python 3.8.13, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/tommaso/Fractal/fractal-server, configfile: pyproject.toml
plugins: asyncio-0.18.3, typeguard-2.13.3, anyio-3.6.2, mock-3.10.0
asyncio: mode=auto
collected 3 items                                                                                                                                                                                                               

tests/test_task_api.py EEE                                                                                                                                                                                                [100%]

============================================================================================================ ERRORS =============================================================================================================
_____________________________________________________________________________________________ ERROR at setup of test_task_get_list ______________________________________________________________________________________________

cls = <class 'fractal_server.app.db.DB'>

    @classmethod
    def engine_sync(cls):
        try:
>           return cls._engine_sync
E           AttributeError: type object 'DB' has no attribute '_engine_sync'

fractal_server/app/db/__init__.py:31: AttributeError

During handling of the above exception, another exception occurred:

event_loop = <_UnixSelectorEventLoop running=False closed=False debug=False>, request = <SubRequest 'db_create_tables' for <Function test_task_get_list>>, kwargs = {'override_settings': None}
setup = <function _wrap_asyncgen.<locals>._asyncgen_fixture_wrapper.<locals>.setup at 0x7f1ae3b8bca0>, finalizer = <function _wrap_asyncgen.<locals>._asyncgen_fixture_wrapper.<locals>.finalizer at 0x7f1ae3b8bd30>

    @functools.wraps(func)
    def _asyncgen_fixture_wrapper(
        event_loop: asyncio.AbstractEventLoop, request: SubRequest, **kwargs: Any
    ) -> _R:
        gen_obj = func(**_add_kwargs(func, kwargs, event_loop, request))
    
        async def setup() -> _R:
            res = await gen_obj.__anext__()
            return res
    
        def finalizer() -> None:
            """Yield again, to finalize."""
    
            async def async_finalizer() -> None:
                try:
                    await gen_obj.__anext__()
                except StopAsyncIteration:
                    pass
                else:
                    msg = "Async generator fixture didn't stop."
                    msg += "Yield only once."
                    raise ValueError(msg)
    
            event_loop.run_until_complete(async_finalizer())
    
>       result = event_loop.run_until_complete(setup())

../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/pytest_asyncio/plugin.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/fractal-server/lib/python3.8/asyncio/base_events.py:616: in run_until_complete
    return future.result()
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/pytest_asyncio/plugin.py:275: in setup
    res = await gen_obj.__anext__()
tests/fixtures_server.py:115: in db_create_tables
    engine = DB.engine_sync()
fractal_server/app/db/__init__.py:33: in engine_sync
    cls.set_db()
fractal_server/app/db/__init__.py:49: in set_db
    cls._engine_async = create_async_engine(
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/ext/asyncio/engine.py:41: in create_async_engine
    sync_engine = _create_engine(*arg, **kw)
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py:309: in warned
    return fn(*args, **kwargs)
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/engine/create.py:560: in create_engine
    dbapi = dialect_cls.dbapi(**dbapi_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'sqlalchemy.dialects.postgresql.asyncpg.PGDialect_asyncpg'>

    @classmethod
    def dbapi(cls):
>       return AsyncAdapt_asyncpg_dbapi(__import__("asyncpg"))
E       ModuleNotFoundError: No module named 'asyncpg'

../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:952: ModuleNotFoundError
_____________________________________________________________________________________________ ERROR at setup of test_collection_api _____________________________________________________________________________________________

cls = <class 'fractal_server.app.db.DB'>

    @classmethod
    def engine_sync(cls):
        try:
>           return cls._engine_sync
E           AttributeError: type object 'DB' has no attribute '_engine_sync'

fractal_server/app/db/__init__.py:31: AttributeError

During handling of the above exception, another exception occurred:

event_loop = <_UnixSelectorEventLoop running=False closed=False debug=False>, request = <SubRequest 'db_create_tables' for <Function test_collection_api>>, kwargs = {'override_settings': None}
setup = <function _wrap_asyncgen.<locals>._asyncgen_fixture_wrapper.<locals>.setup at 0x7f1ae1240c10>, finalizer = <function _wrap_asyncgen.<locals>._asyncgen_fixture_wrapper.<locals>.finalizer at 0x7f1ae1240ca0>

    @functools.wraps(func)
    def _asyncgen_fixture_wrapper(
        event_loop: asyncio.AbstractEventLoop, request: SubRequest, **kwargs: Any
    ) -> _R:
        gen_obj = func(**_add_kwargs(func, kwargs, event_loop, request))
    
        async def setup() -> _R:
            res = await gen_obj.__anext__()
            return res
    
        def finalizer() -> None:
            """Yield again, to finalize."""
    
            async def async_finalizer() -> None:
                try:
                    await gen_obj.__anext__()
                except StopAsyncIteration:
                    pass
                else:
                    msg = "Async generator fixture didn't stop."
                    msg += "Yield only once."
                    raise ValueError(msg)
    
            event_loop.run_until_complete(async_finalizer())
    
>       result = event_loop.run_until_complete(setup())

../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/pytest_asyncio/plugin.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/fractal-server/lib/python3.8/asyncio/base_events.py:616: in run_until_complete
    return future.result()
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/pytest_asyncio/plugin.py:275: in setup
    res = await gen_obj.__anext__()
tests/fixtures_server.py:115: in db_create_tables
    engine = DB.engine_sync()
fractal_server/app/db/__init__.py:33: in engine_sync
    cls.set_db()
fractal_server/app/db/__init__.py:49: in set_db
    cls._engine_async = create_async_engine(
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/ext/asyncio/engine.py:41: in create_async_engine
    sync_engine = _create_engine(*arg, **kw)
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py:309: in warned
    return fn(*args, **kwargs)
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/engine/create.py:560: in create_engine
    dbapi = dialect_cls.dbapi(**dbapi_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'sqlalchemy.dialects.postgresql.asyncpg.PGDialect_asyncpg'>

    @classmethod
    def dbapi(cls):
>       return AsyncAdapt_asyncpg_dbapi(__import__("asyncpg"))
E       ModuleNotFoundError: No module named 'asyncpg'

../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:952: ModuleNotFoundError
____________________________________________________________________________________ ERROR at setup of test_collection_api_invalid_manifest _____________________________________________________________________________________

cls = <class 'fractal_server.app.db.DB'>

    @classmethod
    def engine_sync(cls):
        try:
>           return cls._engine_sync
E           AttributeError: type object 'DB' has no attribute '_engine_sync'

fractal_server/app/db/__init__.py:31: AttributeError

During handling of the above exception, another exception occurred:

event_loop = <_UnixSelectorEventLoop running=False closed=False debug=False>, request = <SubRequest 'db_create_tables' for <Function test_collection_api_invalid_manifest>>, kwargs = {'override_settings': None}
setup = <function _wrap_asyncgen.<locals>._asyncgen_fixture_wrapper.<locals>.setup at 0x7f1ae0bbc670>, finalizer = <function _wrap_asyncgen.<locals>._asyncgen_fixture_wrapper.<locals>.finalizer at 0x7f1ae0bbc700>

    @functools.wraps(func)
    def _asyncgen_fixture_wrapper(
        event_loop: asyncio.AbstractEventLoop, request: SubRequest, **kwargs: Any
    ) -> _R:
        gen_obj = func(**_add_kwargs(func, kwargs, event_loop, request))
    
        async def setup() -> _R:
            res = await gen_obj.__anext__()
            return res
    
        def finalizer() -> None:
            """Yield again, to finalize."""
    
            async def async_finalizer() -> None:
                try:
                    await gen_obj.__anext__()
                except StopAsyncIteration:
                    pass
                else:
                    msg = "Async generator fixture didn't stop."
                    msg += "Yield only once."
                    raise ValueError(msg)
    
            event_loop.run_until_complete(async_finalizer())
    
>       result = event_loop.run_until_complete(setup())

../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/pytest_asyncio/plugin.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/fractal-server/lib/python3.8/asyncio/base_events.py:616: in run_until_complete
    return future.result()
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/pytest_asyncio/plugin.py:275: in setup
    res = await gen_obj.__anext__()
tests/fixtures_server.py:115: in db_create_tables
    engine = DB.engine_sync()
fractal_server/app/db/__init__.py:33: in engine_sync
    cls.set_db()
fractal_server/app/db/__init__.py:49: in set_db
    cls._engine_async = create_async_engine(
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/ext/asyncio/engine.py:41: in create_async_engine
    sync_engine = _create_engine(*arg, **kw)
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py:309: in warned
    return fn(*args, **kwargs)
../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/engine/create.py:560: in create_engine
    dbapi = dialect_cls.dbapi(**dbapi_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'sqlalchemy.dialects.postgresql.asyncpg.PGDialect_asyncpg'>

    @classmethod
    def dbapi(cls):
>       return AsyncAdapt_asyncpg_dbapi(__import__("asyncpg"))
E       ModuleNotFoundError: No module named 'asyncpg'

../../miniconda3/envs/fractal-server/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:952: ModuleNotFoundError
```

--- a/fractal_server/tasks/collection.py
+++ b/fractal_server/tasks/collection.py
@@ -47,7 +47,7 @@ def get_python_interpreter(version: Optional[str] = None) -> str:
 
     Raises:
         ValueError:
-            If the python version requestsed is not available on the host.
+            If the python version requested is not available on the host.
 
     Returns:
         interpreter:

--- a/fractal_server/tasks/collection.py
+++ b/fractal_server/tasks/collection.py
@@ -10,6 +10,7 @@
 # Zurich.
 import json
 import logging
+import shutil
 import sys
 from io import IOBase
 from pathlib import Path
@@ -34,6 +35,34 @@ from ..utils import set_logger
 
 class TaskCollectionError(RuntimeError):
     pass
+
+
+def get_python_interpreter(version: Optional[str] = None) -> str:
+    """
+    Return the path to the python interpreter
+
+    Args:
+        version:
+            Python version
+
+    Raises:
+        ValueError:
+            If the python version requestsed is not available on the host.
+
+    Returns:
+        interpreter:
+            string representing the python executable or its path
+    """
+    if version:
+        interpreter = shutil.which(f"python{version}")
+        if not interpreter:
+            raise ValueError(
+                f"Python version {version} not available on host."
+            )
+    else:
+        interpreter = sys.executable
+
+    return interpreter
 
 
 def get_absolute_venv_path(venv_path: Path) -> Path:
@@ -168,11 +197,7 @@ async def download_package(
     """
     Download package to destination
     """
-    interpreter = (
-        f"python{task_pkg.python_version}"
-        if task_pkg.python_version
-        else sys.executable
-    )
+    interpreter = get_python_interpreter(version=task_pkg.python_version)
     pip = f"{interpreter} -m pip"
     cmd = (
         f"{pip} download --no-deps {task_pkg.pip_package_version} "
@@ -364,10 +389,7 @@ async def _init_venv(
     python_bin : Path
         path to python interpreter
     """
-    if python_version:
-        interpreter = f"python{python_version}"
-    else:
-        interpreter = sys.executable
+    interpreter = get_python_interpreter(version=python_version)
     await execute_command(
         cwd=path, command=f"{interpreter} -m venv venv", logger_name="fractal"
     )

--- a/tests/fixtures_server.py
+++ b/tests/fixtures_server.py
@@ -16,6 +16,7 @@ from dataclasses import field
 from pathlib import Path
 from typing import Any
 from typing import AsyncGenerator
+from typing import Dict
 from typing import List
 from typing import Optional
 from uuid import uuid4
@@ -30,8 +31,12 @@ from fractal_server.config import get_settings
 from fractal_server.config import Settings
 from fractal_server.syringe import Inject
 
+try:
+    import asyncpg  # noqa: F401
 
-DB_ENGINE = "postgres"
+    DB_ENGINE = "postgres"
+except ModuleNotFoundError:
+    DB_ENGINE = "sqlite"
 
 
 def get_patched_settings(temp_path: Path):
@@ -164,21 +169,24 @@ async def MockCurrentUser(app, db):
         """
 
         name: str = "User Name"
+        user_kwargs: Optional[Dict[str, Any]] = None
         scopes: Optional[List[str]] = field(
             default_factory=lambda: ["project"]
         )
         email: Optional[str] = field(
             default_factory=lambda: f"{uuid4()}@exact-lab.it"
         )
-        persist: Optional[bool] = False
+        persist: Optional[bool] = True
 
         def _create_user(self):
-            self.user = User(
-                name=self.name,
+            defaults = dict(
                 email=self.email,
                 hashed_password="fake_hashed_password",
                 slurm_user="slurm_user",
             )
+            if self.user_kwargs:
+                defaults.update(self.user_kwargs)
+            self.user = User(name=self.name, **defaults)
 
         def current_active_user_override(self):
             def __current_active_user_override():

--- a/tests/test_full_workflow.py
+++ b/tests/test_full_workflow.py
@@ -39,6 +39,7 @@ async def collect_tasks(MockCurrentUser, client, dummy_task_package):
     return task_list
 
 
+@pytest.mark.slow
 async def test_full_workflow(
     client,
     MockCurrentUser,

--- a/tests/test_task_api.py
+++ b/tests/test_task_api.py
@@ -4,6 +4,7 @@ from shutil import which as shutil_which
 import pytest
 from devtools import debug
 
+from .fixtures_tasks import execute_command
 from fractal_server.app.api.v1.task import _background_collect_pip
 from fractal_server.app.api.v1.task import _TaskCollectPip
 from fractal_server.app.api.v1.task import create_package_dir_pip
@@ -176,6 +177,10 @@ async def test_collection_api(
         full_path = settings.FRACTAL_ROOT / venv_path
         assert get_collection_path(full_path).exists()
         assert get_log_path(full_path).exists()
+        if python_version:
+            python_bin = data["task_list"][0]["command"].split()[0]
+            version = await execute_command(f"{python_bin} --version")
+            assert python_version in version
 
         # collect again
         res = await client.post(

--- a/tests/test_task_api.py
+++ b/tests/test_task_api.py
@@ -1,4 +1,5 @@
 from pathlib import Path
+from shutil import which as shutil_which
 
 import pytest
 from devtools import debug
@@ -101,7 +102,16 @@ async def test_background_collection_failure(db, dummy_task_package):
 
 @pytest.mark.parametrize(
     ("slurm_user", "python_version"),
-    [("user00", None), ("user01", None)],
+    [
+        ("user00", None),
+        pytest.param(
+            "user01",
+            "3.10",
+            marks=pytest.mark.skipif(
+                not shutil_which("python3.10"), reason="No python3.10 on host"
+            ),
+        ),
+    ],
 )
 async def test_collection_api(
     client, dummy_task_package, MockCurrentUser, slurm_user, python_version

--- a/tests/test_task_api.py
+++ b/tests/test_task_api.py
@@ -99,7 +99,13 @@ async def test_background_collection_failure(db, dummy_task_package):
     assert not venv_path.exists()
 
 
-async def test_collection_api(client, dummy_task_package, MockCurrentUser):
+@pytest.mark.parametrize(
+    ("slurm_user", "python_version"),
+    [("user00", None), ("user01", None)],
+)
+async def test_collection_api(
+    client, dummy_task_package, MockCurrentUser, slurm_user, python_version
+):
     """
     GIVEN a package in a format that `pip` understands
     WHEN the api to collect tasks from that package is called
@@ -113,9 +119,11 @@ async def test_collection_api(client, dummy_task_package, MockCurrentUser):
     """
     PREFIX = "/api/v1/task"
 
-    task_collection = dict(package=dummy_task_package.as_posix())
+    task_collection = dict(
+        package=dummy_task_package.as_posix(), python_version=python_version
+    )
 
-    async with MockCurrentUser(persist=True):
+    async with MockCurrentUser(user_kwargs=dict(slurm_user=slurm_user)):
         # NOTE: collecting private tasks so that they are assigned to user and
         # written in a non-default folder. Bypass for non stateless
         # FRACTAL_ROOT in test suite.

--- a/tests/test_task_collection_pypi.py
+++ b/tests/test_task_collection_pypi.py
@@ -74,8 +74,8 @@ async def test_init_venv(tmp_path, python_version):
     assert (venv_path / "venv/bin/pip").exists()
     assert python_bin.exists()
     assert python_bin == venv_path / "venv/bin/python"
-    version = await execute_command(f"{python_bin} --version")
     if python_version:
+        version = await execute_command(f"{python_bin} --version")
         assert python_version in version
 
 

--- a/tests/test_task_collection_pypi.py
+++ b/tests/test_task_collection_pypi.py
@@ -4,6 +4,7 @@ from pathlib import Path
 import pytest
 from devtools import debug
 
+from .fixtures_tasks import execute_command
 from fractal_server.config import get_settings
 from fractal_server.syringe import Inject
 from fractal_server.tasks.collection import _init_venv
@@ -47,7 +48,8 @@ def test_task_collect_model(dummy_task_package):
     assert tc.package_path == dummy_task_package
 
 
-async def test_init_venv(tmp_path):
+@pytest.mark.parametrize("python_version", [None, "3.10"])
+async def test_init_venv(tmp_path, python_version):
     """
     GIVEN a path and a python version
     WHEN _init_venv() is called
@@ -57,7 +59,14 @@ async def test_init_venv(tmp_path):
     venv_path.mkdir(exist_ok=True, parents=True)
     logger_name = "fractal"
 
-    python_bin = await _init_venv(path=venv_path, logger_name=logger_name)
+    try:
+        python_bin = await _init_venv(
+            path=venv_path,
+            logger_name=logger_name,
+            python_version=python_version,
+        )
+    except ValueError as e:
+        pytest.xfail(reason=str(e))
 
     assert venv_path.exists()
     assert (venv_path / "venv").exists()
@@ -65,6 +74,9 @@ async def test_init_venv(tmp_path):
     assert (venv_path / "venv/bin/pip").exists()
     assert python_bin.exists()
     assert python_bin == venv_path / "venv/bin/python"
+    version = await execute_command(f"{python_bin} --version")
+    if python_version:
+        assert python_version in version
 
 
 async def test_pip_install(tmp_path):

